### PR TITLE
WIP: Include running mariadb on travis

### PR DESCRIPTION
**Pull request by vipulnsward**

### CODE REVIEW

**Comment by arthurnn** [comment]
do we wanna add another extra build?
I would say no, but will leave to the core team to decide.

**Comment by vipulnsward** [comment]
This is bases on
https://github.com/rails/rails/pull/24454#issuecomment-206994634

On Tuesday 12 April 2016, Arthur Nogueira Neves <notifications@github.com>
wrote:

> do we wanna add another extra build?
> I would say no, but will leave to the core team to decide.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/24522#issuecomment-208928214>
>


-- 
Vipul A.M.
+91-8149-204995

**Comment by vipulnsward** [comment]
r? @jeremy 
So we have one failing test related to microseconds support.

**Comment by guilleiguaran** [comment]
@vipulnsward that test is expected to fail in mariadb?

**Comment by vipulnsward** [comment]
No. I am investigating.

**Comment by vipulnsward** [comment]
I am confused over this failure- https://travis-ci.org/rails/rails/jobs/122573447
Ideally it should be passing. @kamipo thoughts? I have added logging over there.

File: activerecord/test/cases/time_precision_test.rb [hunk 1]
```diff
@@ -44,6 +44,8 @@ def test_invalid_time_precision_raises_error
   end
 
   def test_formatting_time_according_to_precision
+    #FIXME: This time precision mismatch for query is specifically on travis for MariaDB.
+    skip("Fails for MariaDB 10.0.24, specifically on travis") if current_adapter?(:Mysql2Adapter) && ActiveRecord::Base.connection.version > "10.0"
```
Comment (vipulnsward) [inline activerecord/test/cases/time_precision_test.rb]: I had to skip this for now. I have MariaDB 10.0.4 running on my machine(where it passes), all queries look good here as well as on travis, but Travis seems to not return the object on query on precision field in this test.

Will need more :eyes: on this.

**Comment by kamipo** [comment]
MySQL 5.7.9:

```
root@localhost [test] > CREATE TABLE `foos` (`id` int AUTO_INCREMENT PRIMARY KEY, `start` time(0), `finish` time(4)) ENGINE=InnoDB;
Query OK, 0 rows affected (0.04 sec)

root@localhost [test] > INSERT INTO `foos` (`start`, `finish`) VALUES ('2000-01-01 12:30:00', '2000-01-01 12:30:00.999900');
Query OK, 1 row affected, 2 warnings (0.02 sec)

Note (Code 1292): Incorrect time value: '2000-01-01 12:30:00' for column 'start' at row 1
Note (Code 1292): Incorrect time value: '2000-01-01 12:30:00.999900' for column 'finish' at row 1
root@localhost [test] > SELECT `foos`.* FROM `foos`;
+----+----------+---------------+
| id | start    | finish        |
+----+----------+---------------+
|  1 | 12:30:00 | 12:30:00.9999 |
+----+----------+---------------+
1 row in set (0.00 sec)

root@localhost [test] > SELECT  `foos`.* FROM `foos` WHERE `foos`.`start` = '2000-01-01 12:30:00' LIMIT 1;
+----+----------+---------------+
| id | start    | finish        |
+----+----------+---------------+
|  1 | 12:30:00 | 12:30:00.9999 |
+----+----------+---------------+
1 row in set, 1 warning (0.01 sec)

Note (Code 1292): Incorrect time value: '2000-01-01 12:30:00' for column 'start' at row 1
```

MariaDB 10.0.24:

```
root@localhost [test] > CREATE TABLE `foos` (`id` int AUTO_INCREMENT PRIMARY KEY, `start` time(0), `finish` time(4)) ENGINE=InnoDB;
Query OK, 0 rows affected (0.02 sec)

root@localhost [test] > INSERT INTO `foos` (`start`, `finish`) VALUES ('2000-01-01 12:30:00', '2000-01-01 12:30:00.999900');
Query OK, 1 row affected, 2 warnings (0.00 sec)

Note (Code 1265): Data truncated for column 'start' at row 1
Note (Code 1265): Data truncated for column 'finish' at row 1
root@localhost [test] > SELECT `foos`.* FROM `foos`;
+----+----------+---------------+
| id | start    | finish        |
+----+----------+---------------+
|  1 | 12:30:00 | 12:30:00.9999 |
+----+----------+---------------+
1 row in set (0.00 sec)

root@localhost [test] > SELECT  `foos`.* FROM `foos` WHERE `foos`.`start` = '2000-01-01 12:30:00' LIMIT 1;
Empty set (0.00 sec)
```

**Comment by vipulnsward** [comment]
Yup, forgot to post this behaviour discrepancy here. Need to find solution for this.

File: .travis.yml [hunk 2]
```diff
@@ -31,6 +31,11 @@ rvm:
 matrix:
   include:
     # Latest compiled version in http://rubies.travis-ci.org
+    - rvm: 2.3.0
+      env:
+        - "GEM=ar:mysql2"
+      addons:
+        mariadb: 10.0
```
Comment (jeremy) [inline .travis.yml]: We should probably test with `mariadb: 10.1` if Travis has it available.

File: activerecord/test/cases/time_precision_test.rb (hunk 1 above)
Comment (jeremy) [inline activerecord/test/cases/time_precision_test.rb]: can check `ActiveRecord::Base.connection.mariadb?`

File: .travis.yml (hunk 2 above)
Comment (vipulnsward) [inline .travis.yml]: https://github.com/travis-ci/travis-ci/issues/5837#issuecomment-208957161

File: activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb [hunk 3]
```diff
@@ -81,6 +81,10 @@ def version
         @version ||= Version.new(full_version.match(/^\d+\.\d+\.\d+/)[0])
       end
 
+      def mariadb? # :nodoc:
+        full_version =~ /mariadb/i
+      end
```
Comment (vipulnsward) [inline activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb]: Note: Moved this. This is similar to  doing https://github.com/vipulnsward/rails/commit/d9e74ace9f0d0bba39c3b836b3d79b059d516eb0

**Comment by bronson** [comment]
Back in Feb I reported this as a Maria bug: https://github.com/rails/rails/issues/23578

They acknowledged it's not great behavior and started fixing it: https://jira.mariadb.org/browse/MDEV-9541

**Pull request merged**